proxytime/proxytime.go

```diff
@@ -89,25 +90,64 @@ func (tm *Time[D]) Tick(d D) {
 	tm.fraction = D(rem)
 }
 
-// FastForwardTo sets the time to the specified Unix timestamp if it is in the
-// future, returning the integer and fraction number of seconds by which the
-// time was advanced. The fraction is always denominated in [Time.Rate].
-func (tm *Time[D]) FastForwardTo(to uint64) (uint64, FractionalSecond[D]) {
-	if to <= tm.seconds {
+// FastForwardTo sets the time to the specified Unix timestamp and fraction of a
+// second, if it is in the future, returning the integer and fraction number of
+// seconds by which the time was advanced. Both the incoming argument and the
+// returned fraction are denominated in [Time.Rate].
+func (tm *Time[D]) FastForwardTo(to uint64, toFrac D) (uint64, FractionalSecond[D]) {
+	to += uint64(toFrac / tm.hertz)
+	toFrac %= tm.hertz
+
+	if !tm.isFuture(to, toFrac) {
 		return 0, FractionalSecond[D]{0, tm.hertz}
 	}
 
-	sec := to - tm.seconds
-	var frac D
-	if tm.fraction > 0 {
-		frac = tm.hertz - tm.fraction
-		sec--
+	ffSec := to - tm.seconds
+	ffFrac := FractionalSecond[D]{
+		Denominator: tm.hertz,
+	}
+
+	switch cmp.Compare(tm.fraction, toFrac) {
+	case -1:
+		ffFrac.Numerator = toFrac - tm.fraction
+	case 0:
+		_ = ffFrac // coverage visualisation
+	case 1:
+		ffFrac.Numerator = tm.hertz - (tm.fraction - toFrac)
+		ffSec--
 	}
 
 	tm.seconds = to
-	tm.fraction = 0
+	tm.fraction = toFrac
 
-	return sec, FractionalSecond[D]{frac, tm.hertz}
+	return ffSec, ffFrac
+}
+
+func (tm *Time[D]) isFuture(sec uint64, num D) bool {
+	switch cmp.Compare(sec, tm.seconds) {
+	case -1:
+		return false
+	case 1:
+		return true
+	default:
+		return num > tm.fraction
+	}
+}
+
+var errGtSecond = errors.New("> 1s")
+
+// ConvertMilliseconds returns `ms`, which MUST be <=1000, as a fraction of a
+// second, denominated in [Time.Rate]. If [Time.Rate] is not a multiple of 1000,
+// the returned fraction will be rounded down.
+func (tm *Time[D]) ConvertMilliseconds(ms uint16) (FractionalSecond[D], error) {
+	frac := FractionalSecond[D]{
+		Denominator: tm.hertz,
+	}
+	if ms > 1000 {
+		return frac, fmt.Errorf("%d milliseconds %w", ms, errGtSecond)
+	}
+	frac.Numerator, _, _ = intmath.MulDiv(D(ms), tm.hertz, 1000)
```

Comment:
Don't we typically enforce that `Numerator < Denominator`?

> ```A FractionalSecond represents a sub-second duration of time.```

But this could result in a full-second duration of time.

I thought that the invariant on this function would be:
> `ms`, which MUST be <1000

Rather than what it is:
> `ms`, which MUST be <=1000